code/modules/flufftext/Hallucination.dm

```diff
@@ -242,34 +251,64 @@ GLOBAL_LIST_INIT(hallucination_list, list(
 		target.Paralyze(100)
 		target.visible_message("<span class='danger'>[target] flails around wildly.</span>","<span class='userdanger'>[name] pounces on you!</span>")
 
+// The numbers of seconds it takes to get to each stage of the xeno attack choreography
+#define XENO_ATTACK_STAGE_LEAP_AT_TARGET 1
+#define XENO_ATTACK_STAGE_LEAP_AT_PUMP 2
+#define XENO_ATTACK_STAGE_CLIMB 3
+#define XENO_ATTACK_STAGE_FINISH 6
+
+/// Xeno crawls from nearby vent,jumps at you, and goes back in
 /datum/hallucination/xeno_attack
-	//Xeno crawls from nearby vent,jumps at you, and goes back in
-	var/obj/machinery/atmospherics/components/unary/vent_pump/pump = null
+	var/turf/pump_location = null
 	var/obj/effect/hallucination/simple/xeno/xeno = null
+	var/time_processing = 0
+	var/stage = XENO_ATTACK_STAGE_LEAP_AT_TARGET
 
 /datum/hallucination/xeno_attack/New(mob/living/carbon/C, forced = TRUE)
-	set waitfor = FALSE
 	..()
 	for(var/obj/machinery/atmospherics/components/unary/vent_pump/U in orange(7,target))
 		if(!U.welded)
-			pump = U
+			pump_location = get_turf(U)
 			break
-	if(pump)
-		feedback_details += "Vent Coords: [pump.x],[pump.y],[pump.z]"
-		xeno = new(pump.loc,target)
-		sleep(10)
-		xeno.update_icon("alienh_leap",'icons/mob/alienleap.dmi',-32,-32)
-		xeno.throw_at(target,7,1, xeno, FALSE, TRUE)
-		sleep(10)
-		xeno.update_icon("alienh_leap",'icons/mob/alienleap.dmi',-32,-32)
-		xeno.throw_at(pump,7,1, xeno, FALSE, TRUE)
-		sleep(10)
-		var/xeno_name = xeno.name
-		to_chat(target, "<span class='notice'>[xeno_name] begins climbing into the ventilation system...</span>")
-		sleep(30)
-		qdel(xeno)
-		to_chat(target, "<span class='notice'>[xeno_name] scrambles into the ventilation ducts!</span>")
-	qdel(src)
+
+	if(pump_location)
+		feedback_details += "Vent Coords: [pump_location.x],[pump_location.y],[pump_location.z]"
+		xeno = new(pump_location, target)
+		START_PROCESSING(SSfastprocess, src)
+	else
+		qdel(src)
+
+/datum/hallucination/xeno_attack/process(delta_time)
+	time_processing += delta_time
+
+	if (time_processing >= stage)
```

Comment:
Early return, also keep in mind this is a float comparison

Reply:
Float *equality* is unsafe--checking order (`>` and `<`) is okay.

Reply:
I bring it up because I ran into issues with timing in a loop like this with #55860, `counter >=  delay` was fucking me up, but I was doing some dumb stuff along side it, so it may not actually be an issue for you

Reply:
If these kinds of comparisons are unreliable, we have worse problem 😛